fix(VAutocomplete): divider/subheader not rendering

func0der opened the pull request:
Original PR: https://github.com/vuetifyjs/vuetify/pull/15728

## Description
This is just an updated version of https://github.com/vuetifyjs/vuetify/pull/15728, since the functionality is still missing and the PR was closed as stale.
@nekosaur I hope this was okay?

You are welcome to update and fix accordingly.

I am not 100% in the Vuetify game and I might have made mistakes merging `master`. I am especially insecure about the changes in `VTreeView`, since the component was not present when the original PR was created.

## Markup:
As described here: https://vuetifyjs.com/en/getting-started/contributing/#for-docs-language

<!-- Paste your FULL packages/vuetify/dev/Playground.vue here --->
```vue
<template>
  <v-card class="mx-auto" max-width="300">
    <v-list :items="items" />
    <v-select :items="items" />
    <v-autocomplete :items="items" />
    <v-treeview :items="items" />
  </v-card>
</template>

<script>
  export default {
    data: () => ({
      items: [
        {
          title: 'Group (Dropdown in List, old behavior)',
          value: 'group',
          children: [
            {
              title: 'Child',
              subtitle: 'Subtitle',
              value: 'child',
            },
          ],
        },
        {
          props: {
            subheader: true,
            divider: true,
          },
          title: 'Group #1 (subheader in list, additional new behavior)',
          value: 'foo',
          children: [
            {
              title: 'Item #1',
              value: 1,
            },
            {
              title: 'Item #2',
              value: 2,
            },
            {
              title: 'Item #3',
              value: 3,
            },
          ],
        },
        {
          props: {
            subheader: true,
          },
          title: 'Group #2 (choose to have no divider)',
          children: [
            {
              title: 'Item #4',
              value: 4,
            },
            {
              title: 'Item #5',
              value: 5,
            },
            {
              title: 'Item #6',
              value: 6,
            },
          ],
        },
      ],
    }),
  }
</script>

```

--- update 1:

Updated the playground items array to better reflect how the items diff from each other.

func0der commented:
Known issues for now are:
* Failing test because the rending of groups is somehow broken (have not found out why yet)
* The syntax to define subheaders and dividers is unique to this usecase and different from the way of VListItem ,which would result in a breaking change and not a feature
* There might still be unnecessary code in the like the `!slots.default` lines in `VListItem` that I can not quite place in the context of this issue

func0der commented on packages/vuetify/src/components/VList/VListChildren.tsx:
```diff
@@ -43,65 +45,53 @@ export const VListChildren = genericComponent<new <T extends InternalListItem>(
+            }}
+          </VListGroup>
+        ) : (
+          slots.item ? slots.item({ item: internalItem, props: itemProps, index }) : (
```
is this kind of change breaking or do you not need to use ALL available properties when redefining slots?

func0der replied:
My question was about the definition of a custom slot.
So for example, if user has this in their template:

```
<template #item="{ props }">

</template>
```

Would adding new available parameters for the slot here, break their code.

But since we are not changing the naming of `props`, it should work fine :)

func0der commented:
@yuwu9145 @johnleider @KaelWD 
Any chance we can get a review on this, so I can work on possible change requests and we can continue to work on this? :)

johnleider commented on packages/vuetify/src/components/VList/VListItem.tsx:
```diff
@@ -294,13 +294,13 @@ export const VListItem = genericComponent<VListItemSlots>()({
 
           <div class="v-list-item__content" data-no-activator="">
-            { hasTitle && (
+            { hasTitle && !slots.default && (
```
What is the reason for this change?

func0der replied:
Actually. I have no real clue. This is something I just copied over from the other PR. It came with this commit, which I think is not really related to this change: https://github.com/vuetifyjs/vuetify/commit/e6792c1b79b0e0a5cab35db6720dd5b62dfc2d0a

If it is not needed or is bad practice, I will remove it.

johnleider commented on packages/vuetify/src/components/VList/VListItem.tsx:
```diff
@@ -294,13 +294,13 @@ export const VListItem = genericComponent<VListItemSlots>()({
             )}
 
-            { hasSubtitle && (
+            { hasSubtitle && !slots.default && (
```
What is the reason for this change?

func0der replied:
Actually. I have no real clue. This is something I just copied over from the other PR. It came with this commit, which I think is not really related to this change: https://github.com/vuetifyjs/vuetify/commit/e6792c1b79b0e0a5cab35db6720dd5b62dfc2d0a

If it is not needed or is bad practice, I will remove it.

johnleider commented on packages/vuetify/src/components/VSelect/VSelect.tsx:
```diff
@@ -85,9 +86,12 @@ export const makeSelectProps = propsFactory({
     default: '$vuetify.noDataText',
   },
   openOnClear: Boolean,
+  readonly: Boolean,
```
Is this related?

func0der replied:
No. It is a relic of a commit I have removed, since it was unrelated to the issue at hand: https://github.com/vuetifyjs/vuetify/commit/6d6c1dc43eb3d9018420ebdbee57ee2b8ff51e42

I will remove it.

johnleider commented on packages/vuetify/src/composables/list-items.ts:
```diff
@@ -63,7 +63,7 @@ export function transformItem (props: Omit<ItemProps, 'items'>, item: any): List
     ? typeof item === 'object' && item != null && !Array.isArray(item)
       ? 'children' in item
-        ? omit(item, ['children'])
+        ? omit(item as any, ['children'])
```
This use of "any" here is probably a red flag for @KaelWD

func0der replied:
I can remove it, I think, since `transformItem` already types it as `any`.

johnleider reviewed (changes requested):
What is the reasoning from switching from InternalListItem to ListItem?

func0der commented:
@nekosaur Would love for you to chip in on some of those review comments/question, if you can :)

func0der commented:
> What is the reasoning from switching from InternalListItem to ListItem?

I think that the `InternalListItem` was more of a helper construct inside the component and not something that was actually used.
Since the whole list flattening is now inside `list-items.ts` it is not needed anymore. Inside `list-items.ts` the type is not called `FlatListItem`.

gduliscouet-ubitransport commented on packages/vuetify/src/components/VSelect/VSelect.tsx:
```diff
@@ -87,7 +88,9 @@ export const makeSelectProps = propsFactory({
 
-  ...makeItemsProps({ itemChildren: false }),
+  ...makeItemsProps({
+    itemProps: () => (item: any) => ({ subheader: item.subheader, divider: item.divider }),
```
I tested the branch with the provided markup and the dividers are missing for the VSelect (and VAutocomplete). It seems like this is expecting `subheader` and `divider` to be at the root, but it is not the case.

![image](https://github.com/vuetifyjs/vuetify/assets/101416921/86b819e8-b55e-4f6e-8b89-87d818b105ec)


Looking into `item.props?.subheader` works. It also works like that:

```suggestion
    itemProps: 'props',
```

![image](https://github.com/vuetifyjs/vuetify/assets/101416921/b5f8348e-aaf2-4130-a0b1-7c1536d51e4a)

gduliscouet-ubitransport replied:
Unless the markup you provided is not the good one @func0der ? I see on the [original PR](https://github.com/vuetifyjs/vuetify/pull/15728) that `subheader` and `divider` were at the root of the item

func0der replied:
Thanks for the pointer. 
I could not make out any occurence where this change would be needed.
I guess, it is just a leftover from the time this features was experimented with.

I changed this to have only the `itemChildren` removed.

gduliscouet-ubitransport replied:
It looks good with your change @func0der, thank you !

gduliscouet-ubitransport commented:
We are quite interested in this feature and I would love to help moving this PR forward. But it is, I think, too much to handle safely by people unfamilliar with the Vuetify codebase. Would a core member be able to help on this ? @johnleider @KaelWD ?

johnleider commented:
> We are quite interested in this feature and I would love to help moving this PR forward. But it is, I think, too much to handle safely by people unfamilliar with the Vuetify codebase. Would a core member be able to help on this ? @johnleider @KaelWD ?

I've asked @nekosaur if he could take a look at some point when he has time.

func0der commented:
Sorry for the late reaction. I have addressed your remarks. :) 

> But it is, I think, too much to handle safely by people unfamilliar with the Vuetify codebase. Would a core member be able to help on this ?

I'll try my best anyway so they only have to check it afterwards. After all it is just an adoption of @nekosaur's code :)

egeersoz commented:
Thank you for this PR. This functionality is really important IMO and I hope we get it back soon!

egeersoz commented:
Just wanted to inquire about the status of this PR. Thanks!

egeersoz commented:
Any chance this can be reviewed and merged? We had to add a lot of conditionals inside our slots to make it work, and miss the way it worked seamlessly in v2.
